### PR TITLE
[feat] sync adascale from internal repo, support add_param_group

### LINKED ISSUE #224: [feat] AdaScale: supporting separate scale vs. world_size
## 🚀 Feature
@mikerabbat and @erikwijmans already worked out the math and have a version of this. This PR is keep track of porting that into this code base. The idea is that user can choose different `scale` value and `world_size`. Currently scale equals to `world_size * grad_acumu`. Per-GPU batch size is assume to be equal to scale-1 batch size.

However, it is possible that scale-1 batch size != Per-GPU batch size. In that case, scale != world_size * grad_accum.

## Motivation

This will be more flexible to the users.

## Pitch

See above.

## Alternatives

API-wise, we may want to specify base_scale_1_batch_size instead of the scale value. Mike thinks it is the same. I think base_scale_1_batch_size is better and easier to understand.

## Additional context

n/a

### LINKED ISSUE #269: [feat] adascale to support add_param_group
This optimizer API is sometime used. Better to support it.

### PR DESCRIPTION
**Pull request by min-xu-ai**
- [x] added add_param_group API, unit test added
- [x] added code to support scale != world_size, unit test added
- [x] added code to unhook the hooks, unit test added
- [x] removed `scale` param from the gain() function since it is not correct to do it there.
- [x] changed to not mutate the gradient in-place in the grad-accumulation case, unit test updated
- [x] added experimental feature for debias_ewma or not, unit test added
- [x] added experimental code to adjust gradient accumulation steps during the training (`set_num_gradients_to_accumulate`), unit test added
- [x] added checking code for proper API calling places
- [x] rename internal APIs to make them internal
- [x] update changelog

testing: unit tests

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes #269.
Fixes #224.

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/optim/adascale.py [hunk 1]
```diff
@@ -322,33 +428,38 @@ def _final_callback(self) -> None:
...
+        # total_grad_sqr is \norm{\bar{g}_t}^2
         S = self._scale
-        grad_var = local_grad_sqr / (S - 1) - total_grad_sqr * S / (S - 1)
+        cN = self._world_size * self._num_grads_to_accum
+        grad_var = local_grad_sqr * (S / cN) / (cN - 1) - total_grad_sqr * S / (cN - 1)
         grad_sqr = total_grad_sqr - grad_var / S
         grad_var = np.maximum(grad_var, 1e-6)
         grad_sqr = np.maximum(grad_sqr, 0.0)
-        theta = self._smoothing
+        theta = self.smoothing
```
Comment (blefaudeux) [inline fairscale/optim/adascale.py:457]: is `theta` really needed here ?

**Review by blefaudeux — APPROVED** [review]
LGTM, terrific that there are so many new tests and everything looks very clean indeed, but I'm missing a lot of context and small errors would be hard to catch, could be a good idea to wait for Mike ?

File: fairscale/optim/adascale.py (hunk 1 above)
Comment (min-xu-ai) [reply fairscale/optim/adascale.py:457]: Good point! Will simplify it. Thanks for the speedy review! I am not in a hurry to merge. Will wait for Mike. :-)

**Review by mikerabbat — APPROVED** [review]
Looks great. Thanks for writing the additional tests Min!
